File: android_feature_topics/src/main/java/com/app/feature/topics/TopicsViewState.kt
package com.app.feature.topics

import com.app.shared.business.BookmarkState
import com.app.shared.business.TopicState
import com.app.shared.business.TopicsState
import com.app.views.viewstate.TopicViewState

data class TopicsViewState(
    val state: TopicsState,
    val bookmarks: List<BookmarkState>
) {
    val topics: List<TopicViewState> = state.topics.map {
        TopicViewState.Normal(state = it, noBookmarks = noBookmarks(fromTopic = it))
    }

    private fun noBookmarks(fromTopic: TopicState): Int {
        return bookmarks.filter { it.topic?.id == fromTopic.id }.size
    }
}

File: ios_app/ios_feature_topics/TopicsViewState.swift
//
//  TopicsViewState.swift
//  ios_feature_topics
//
//  Created by Gabriel Coman on 29/04/2020.
//  Copyright © 2020 Gabriel Coman. All rights reserved.
//

import Foundation
import ios_views
import RememberShared

struct TopicsViewState {
    private let state: [TopicState]
    private let bookmarks: [BookmarkState]
    
    init(
        state: [TopicState] = [],
        bookmarks: [BookmarkState] = []
    ) {
        self.state = state
        self.bookmarks = bookmarks
    }
    
    var viewStates: [TopicViewState.Normal] {
        return state.map {
            TopicViewState.Normal(
                state: $0,
                noBookmarks: self.getNoBookmarks(fromTopic: $0)
            )
        }
    }
    
    private func getNoBookmarks(fromTopic topic: TopicState) -> Int {
        return bookmarks.filter { $0.topic?.id == topic.id }.count
    }
}
